### PR TITLE
JENKINS-1877 Add ability to specify shut down reason

### PR DESCRIPTION
**Pull request by Austry**
[JENKINS-1877](https://issues.jenkins-ci.org/browse/JENKINS-1877).

Allow specifying a reason for quieting down Jenkins from UI, CLI, and API.

### Proposed changelog entries

* Add the ability to specify a reason for quieting down Jenkins ("Prepare for shutdown")

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change).
- [x] Appropriate autotests

### CODE REVIEW

**Comment by res0nance** [comment]
Test flake, retriggering CI

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (res0nance) [inline core/src/main/java/jenkins/model/Jenkins.java]: IIRC Nullable does nothing but checkfornull ensures that spotbugs forces people to check for null

**Review by jvz — APPROVED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Would be nice to have some Javadoc added to the method

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
                                    @QueryParameter @CheckForNull String reason) throws InterruptedException, IOException {
```

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/lib/layout/main-panel.jelly]: Should it rather go through the Markup formatter so that one can set hyperlinks & Co?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
    @CheckForNull
```

**Review by oleg-nenashev — COMMENTED** [review]
Looks good overall, my comments are rather suggestions (Nullable is NOOP for SpotBugs). Would it make sense to also rename the code to `quietDownReason` so that it it consistent with existing API?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Requesting changes to indicate that I expect suggestions above to be addressed before merging.
All these suggestions are fairly minor, so I am fine if there is a consensus between there reviewers to merge the PR as is. Anyway, it is almost ready to go, thanks @Austry !

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (Austry) [reply core/src/main/resources/lib/layout/main-panel.jelly]: Can you give me a hint how to do it?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
     *      Reason if it was indicated. null otherwise
     *      @since TODO
```

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: Breaking change of non-`@Restricted` API. At a minimum, confirm no uses in the plugin ecosystem.

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (daniel-beck) [reply core/src/main/resources/lib/layout/main-panel.jelly]: `Jenkins.get().getMarkupFormatter().translate(String)`, there might be a convenient utility in `Functions` for this but I'm not sure.

Would consider this out of scope for this one.

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/main-panel.jelly]: Would it make sense to use the opportunity of these parts being worked on to think about the wording?

There are other reasons than a planned restart to quiet down the instance, and since we're now adding an explanation what's going on, the built-in messaging could more generically just state "Jenkins is quieting down" or "Jenkins is suspending execution of further tasks" or something like that.

Thoughts?

**Review by daniel-beck — CHANGES_REQUESTED** [review]
In general, this looks reasonable.

Requesting a fix/investigation of the potential breaking change, the others would like to see considered.

Personally I'd strongly prefer if we get the complete package in one PR; that way, suboptimal design here would not negatively impact the followup change without us being able to detect it.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Austry) [reply core/src/main/java/jenkins/model/Jenkins.java]: Added a method with an old signature for compatibility

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (Austry) [reply core/src/main/resources/lib/layout/main-panel.jelly]: I do not mind and can do it. But a little sorry to lose translations

**Review by res0nance — APPROVED** [review]

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (oleg-nenashev) [reply core/src/main/resources/lib/layout/main-panel.jelly]: Imho can be done in a follow-up, we can create a ticket for it

**Review by oleg-nenashev — APPROVED** [review]
Thanks! Looks good after the fixes

**Comment by daniel-beck** [comment]
Any consideration for

> Personally I'd strongly prefer if we get the complete package in one PR; that way, suboptimal design here would not negatively impact the followup change without us being able to detect it.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
See previous comment. This looks like a risky change as I see nothing that would confirm the new method takes precedence in request routing. In fact, it introduces a web method without actually using it.

If the intention is for this to be purely programmatic API, is should probably be `setQuietDown` or something along those lines, rather than have a dual programmatic API / web method (which IMO is an anti-pattern to begin with).

"Requesting changes" for at a minimum confirming that HTTP requests from existing UI go through the new method, otherwise "part 2" will be dead on arrival. Ideally "part 2" should be integrated into this change already.

**Comment by Austry** [comment]
PR was updated and rebased to fresh master branch.
@daniel-beck , new screen for shut down reason specification added(see screenshots in attaches). It's now available from manage Jenkin's page with ShutdownLink. Any advices about UI implementation are welcome :)

File: core/src/main/resources/jenkins/management/Messages.properties
Comment (timja) [inline core/src/main/resources/jenkins/management/Messages.properties]: Seems unneeded to have text in the wording
```suggestion
ShutdownLink.ShutDownReason_update=Update reason
```

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/ShutdownLink.java]: Lacks permission check. While `#doQuietDown` does, this logs messages before permission is checked.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/ShutdownLink.java]: Lacks permission check. While `#doCancelQuietDown` does, this logs messages before permission is checked.

**Review by daniel-beck — COMMENTED** [review]
Planning to do a more detailed review + test soon, but this stood out.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (res0nance) [inline core/src/main/java/jenkins/model/Jenkins.java]: Probably shouldn't be synchronized, and it might make sense for it to be deprecated

**Review by res0nance — CHANGES_REQUESTED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Austry) [reply core/src/main/java/jenkins/model/Jenkins.java]: Don't wanna break anything. Let it be synchronized.
And I will mark it as deprecated.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java]: I think it's quite reasonable logging before permission check. If user don't have permission - exception will be logged. But in any situation(user have permission or not) we will log that action was `requested`. It can help in problems investigations. But if you insist I can change it.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java]: See https://github.com/jenkinsci/jenkins/pull/4274#discussion_r383155118

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (varyvol) [reply core/src/main/java/jenkins/management/ShutdownLink.java:67]: What is logged if the user doesn't have permissions? Any kind of information that might be private?

**Comment by varyvol** [comment]
@Austry could you please resolve the merge conflict?
@res0nance @daniel-beck @timja could you please review again? It seems that your requests/comments have been attended.

**Review by fqueiruga — APPROVED** [review]
Approving the UI code, it looks solid

**Review by res0nance — APPROVED** [review]

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java:70]: Permission check added

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java:84]: Permission check added

**Comment by Austry** [comment]
@varyvol conflicts resolved

**Review by varyvol — APPROVED** [review]

**Comment by varyvol** [comment]
@daniel-beck @timja could you please dedicate some time to review this? It actually seems nice to have and basically ready to be merged.

**Review by alecharp — APPROVED** [review]

File: core/src/main/resources/jenkins/management/Messages.properties
Comment (timja) [inline core/src/main/resources/jenkins/management/Messages.properties:59]:
```suggestion
ShutdownLink.ShutDownReason_title=Reason
```

**Review by timja — APPROVED** [review]
All looks good, one suggestion on shortening the text, it's always in context that it's a shutdown so I don't think we need to repeat it

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/ShutdownLink.java:45]: Incompatible symbol change. Unsure whether we care but should not go without comment.

If we're changing it, should use the term "quiet down" IMO.

**Comment by daniel-beck** [comment]
> @daniel-beck […] could you please dedicate some time to review this? It actually seems nice to have and basically ready to be merged.

Will try to review it later this week. If I don't get to it this week, feel free to merge.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:3929]:
```suggestion
     * @since TODO
     */
```

(Or perhaps we want this "HTTP API" to be `@Restricted`?)

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:3899]: Why are we adding this method when the intended new endpoint is in the `ShutdownLink` anyway? Or is one intentionally as the UI endpoint and the other as API endpoint?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:2822]: Wouldn't it be cleaner to use the new `quietDownReason` as the only indicator for quiet down state?

E.g.:

- `null`: Not quieting down
- `""`: Quieting down, no reason
- any other value: Quieting down, this reason

(Or a proper value object with a basic, sane API)

Having two independent variables seems to invite inconsistent states.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/ShutdownLink.java:56]: I'm unsure about this UI change.

It makes sense that the management of the quiet down reason is placed in the `ManagementLink` form, but it is confusing to see the link "Prepare for shutdown" when we're already in this state.

I wonder whether we could retain the existing behavior, perhaps with a different label instead of the now not-quite-applicable "Cancel Shutdown", like "Update shutdown preparation"?

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/ShutdownLink.java:76]: Might make sense to log the user here, we do this for restarts.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:3912]:
```suggestion
     * @deprecated since TODO; use {@code #doQuietDown(boolean,int,String)} instead.
```

(Untested Javadoc, should be verified in IDE.)

File: core/src/main/resources/lib/layout/main-panel.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/main-panel.jelly:35]: Markup formatter would be nice, especially in cases when admins would link to a URL showing planned maintenances or something.

```suggestion
          <br/> ${%Shut down reason}: ${app.markupFormatter.translate(app.getQuietDownReason())}
```

That said, probably better off as a future enhancement, as it would need a preview feature for the shutdown reason text field too.

**Review by daniel-beck — DISMISSED** [review]
Requesting a change only for the following:

1. To have the question about the symbol change addressed (an informed "nobody will care" works for me).
2. To have https://github.com/jenkinsci/jenkins/blob/a0c9a062705912c41f0959cac833767ac4ba3cef/core/src/main/resources/jenkins/model/Jenkins/_api.jelly#L40-L49 updated with a note about the new `?reason=bla` query parameter.

Other feedback might be nice to have, but this is basically conditional approval. The PR looks really great! Thanks for the submission and your patience with reviews.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (timja) [reply core/src/main/java/jenkins/management/ShutdownLink.java:45]: not even sure why we need a symbol for a management link?

Comment (daniel-beck) [reply core/src/main/java/jenkins/management/ShutdownLink.java:45]: 🤷 Ask KK.

But might be as simple as having a generic `ExtensionFilter`, configurable within JCasC.

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java:76]: - [ ] Log user

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java:56]: - [ ] Show different text for link to shutdown page

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Austry) [reply core/src/main/java/jenkins/model/Jenkins.java:2822]: I agree with you, will refactor it to one indicator.

- [ ] Refactor to one shutdown state field

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Austry) [reply core/src/main/java/jenkins/model/Jenkins.java:3912]: - [ ] Check proposed javadoc

File: core/src/main/java/jenkins/management/ShutdownLink.java
Comment (Austry) [reply core/src/main/java/jenkins/management/ShutdownLink.java:45]: - [ ] change symbol to "prepareQuietdown"

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (Austry) [reply core/src/main/java/jenkins/model/Jenkins.java:3929]: - [ ] Insert todo

**Comment by Austry** [comment]
- [ ] Update "Restarting Jenkins" info paragraph

**Comment by Austry** [comment]
Thank you for review!
I planning to make changes to address comments in a week or two.

**Comment by alecharp** [comment]
@daniel-beck do you think that your comments were addressed? If so, maybe we can go ahead and mark this PR as `ready-for-merge`. Thanks.

**Comment by Austry** [comment]
@alecharp I plan to fix them in a week. 
If @daniel-beck consider they as not important, then PR can be merged

File: core/src/main/resources/jenkins/model/Jenkins/_api.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/_api.jelly:42]:
```suggestion
    Jenkins will enter into the "quiet down" mode by sending a POST request with optional <code>reason</code> query parameter to <a href="../quietDown">this URL</a>.
    You can also send another request to this URL to update the reason.
```

File: core/src/main/resources/jenkins/model/Jenkins/_api.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/_api.jelly:43]:
```suggestion
    You can cancel this mode by sending a POST request to <a href="../cancelQuietDown">this URL</a>. On environments
```

**Review by daniel-beck — COMMENTED** [review]
Would prefer the outstanding comments be addressed or finally dismissed. Especially collapsing the two fields to one intends to prevent possible overlooked bugs; adding that after we release this seems a bit silly.

File: core/src/main/resources/jenkins/management/ShutdownLink/index.groovy
Comment (daniel-beck) [inline core/src/main/resources/jenkins/management/ShutdownLink/index.groovy:9]:
```suggestion
l.layout(norefresh: true, permission: app.MANAGE, title: my.displayName) {
```

Should be tested with Overall/Manage permission enabled (`Jenkins.MANAGE.enabled = true` + `matrix-auth` plugin for fine-grained permissions + a second user without Overall/Administer).

Comment (Austry) [reply core/src/main/resources/jenkins/management/ShutdownLink/index.groovy]: - [ ] Update permission
Seems like it's working

**Comment by Austry** [comment]
All comments and requested changes are addressed
@daniel-beck can you review it?

**Comment by daniel-beck** [comment]
@Austry Thanks a lot for amending this PR. At a glance this looks great. I plan to give it a trial run tomorrow or Monday so it _should_ be ready towards the next weekly release.

**Comment by MRamonLeon** [comment]
@daniel-beck would you have a gap to look at it?

**Review by MRamonLeon — APPROVED** [review]

**Comment by MarkEWaite** [comment]
@rsandell with the conflicting files and the request for changes, I assumed this was not ready for merge and removed the "ready-for-merge" label.

**Comment by alecharp** [comment]
@Austry could you please review the merge conflict? Thanks.

**Comment by Austry** [comment]
@alecharp Done

**Comment by daniel-beck** [comment]
Looks like I forgot about this PR after my previous comment, I'm really sorry!

I'll take a look this weekend, once the PR build is done. And just to be safe, this should not be taken as blocking this PR.

**Review by daniel-beck — APPROVED** [review]

**Comment by daniel-beck** [comment]
We can probably merge this before the upcoming weekly, WDYT?

**Comment by daniel-beck** [comment]
@timja Thanks for the merge, I feared this didn't make it in when I saw the release being uploaded 😅

**Comment by timja** [comment]
Was last minute :p the release almost had nothing in it

**Pull request merged**
